File: backend/internal/models/submission.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// Submission represents a conversation thread for student-instructor feedback
type Submission struct {
	ID        uuid.UUID  `json:"id" db:"id"`
	ProgramID uuid.UUID  `json:"program_id" db:"program_id"`
	UserID    uuid.UUID  `json:"user_id" db:"user_id"` // Student who created it
	Title     string     `json:"title" db:"title"`
	CreatedAt time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt time.Time  `json:"updated_at" db:"updated_at"`
	DeletedAt *time.Time `json:"deleted_at,omitempty" db:"deleted_at"`
}

// SubmissionMessage represents an individual message in a submission conversation
type SubmissionMessage struct {
	ID           uuid.UUID `json:"id" db:"id"`
	SubmissionID uuid.UUID `json:"submission_id" db:"submission_id"`
	UserID       uuid.UUID `json:"user_id" db:"user_id"` // Author (student or instructor)
	Content      string    `json:"content" db:"content"`
	YouTubeURL   *string   `json:"youtube_url,omitempty" db:"youtube_url"`
	CreatedAt    time.Time `json:"created_at" db:"created_at"`
}

// MessageReadStatus tracks which users have read which messages
type MessageReadStatus struct {
	UserID    uuid.UUID `json:"user_id" db:"user_id"`
	MessageID uuid.UUID `json:"message_id" db:"message_id"`
	ReadAt    time.Time `json:"read_at" db:"read_at"`
}

// SubmissionWithMessages includes submission and all its messages
type SubmissionWithMessages struct {
	Submission Submission          `json:"submission"`
	Messages   []SubmissionMessage `json:"messages"`
}

// SubmissionListItem is used for list views with metadata
type SubmissionListItem struct {
	Submission
	ProgramName     string    `json:"program_name" db:"program_name"`
	StudentName     string    `json:"student_name" db:"student_name"`
	StudentEmail    string    `json:"student_email" db:"student_email"`
	MessageCount    int       `json:"message_count" db:"message_count"`
	UnreadCount     int       `json:"unread_count" db:"unread_count"`
	LastMessageAt   time.Time `json:"last_message_at" db:"last_message_at"`
	LastMessageText string    `json:"last_message_text" db:"last_message_text"`
	LastMessageFrom string    `json:"last_message_from" db:"last_message_from"`
}

// MessageWithAuthor includes message with author details
type MessageWithAuthor struct {
	SubmissionMessage
	AuthorName  string   `json:"author_name" db:"author_name"`
	AuthorEmail string   `json:"author_email" db:"author_email"`
	AuthorRole  UserRole `json:"author_role" db:"author_role"`
	IsRead      bool     `json:"is_read" db:"is_read"` // For current user
}

// UnreadCounts holds unread message counts at various levels
type UnreadCounts struct {
	Total        int            `json:"total"`
	ByProgram    map[string]int `json:"by_program"`
	BySubmission map[string]int `json:"by_submission"`
}

File: app/lib/models/submission.dart
class Submission {
  final String id;
  final String programId;
  final String userId;
  final String title;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? deletedAt;

  Submission({
    required this.id,
    required this.programId,
    required this.userId,
    required this.title,
    required this.createdAt,
    required this.updatedAt,
    this.deletedAt,
  });

  factory Submission.fromJson(Map<String, dynamic> json) {
    return Submission(
      id: json['id'],
      programId: json['program_id'],
      userId: json['user_id'],
      title: json['title'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      deletedAt: json['deleted_at'] != null
          ? DateTime.parse(json['deleted_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'program_id': programId,
      'user_id': userId,
      'title': title,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'deleted_at': deletedAt?.toIso8601String(),
    };
  }
}

class SubmissionMessage {
  final String id;
  final String submissionId;
  final String userId;
  final String content;
  final String? youtubeUrl;
  final DateTime createdAt;

  SubmissionMessage({
    required this.id,
    required this.submissionId,
    required this.userId,
    required this.content,
    this.youtubeUrl,
    required this.createdAt,
  });

  factory SubmissionMessage.fromJson(Map<String, dynamic> json) {
    return SubmissionMessage(
      id: json['id'],
      submissionId: json['submission_id'],
      userId: json['user_id'],
      content: json['content'],
      youtubeUrl: json['youtube_url'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'submission_id': submissionId,
      'user_id': userId,
      'content': content,
      'youtube_url': youtubeUrl,
      'created_at': createdAt.toIso8601String(),
    };
  }
}

class MessageWithAuthor {
  final String id;
  final String submissionId;
  final String userId;
  final String content;
  final String? youtubeUrl;
  final DateTime createdAt;
  final String authorName;
  final String authorEmail;
  final String authorRole;
  final bool isRead;

  MessageWithAuthor({
    required this.id,
    required this.submissionId,
    required this.userId,
    required this.content,
    this.youtubeUrl,
    required this.createdAt,
    required this.authorName,
    required this.authorEmail,
    required this.authorRole,
    required this.isRead,
  });

  factory MessageWithAuthor.fromJson(Map<String, dynamic> json) {
    return MessageWithAuthor(
      id: json['id'],
      submissionId: json['submission_id'],
      userId: json['user_id'],
      content: json['content'],
      youtubeUrl: json['youtube_url'],
      createdAt: DateTime.parse(json['created_at']),
      authorName: json['author_name'],
      authorEmail: json['author_email'],
      authorRole: json['author_role'],
      isRead: json['is_read'] ?? false,
    );
  }
}

class SubmissionListItem {
  final String id;
  final String programId;
  final String userId;
  final String title;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String programName;
  final String studentName;
  final String studentEmail;
  final int messageCount;
  final int unreadCount;
  final DateTime lastMessageAt;
  final String lastMessageText;
  final String lastMessageFrom;

  SubmissionListItem({
    required this.id,
    required this.programId,
    required this.userId,
    required this.title,
    required this.createdAt,
    required this.updatedAt,
    required this.programName,
    required this.studentName,
    required this.studentEmail,
    required this.messageCount,
    required this.unreadCount,
    required this.lastMessageAt,
    required this.lastMessageText,
    required this.lastMessageFrom,
  });

  factory SubmissionListItem.fromJson(Map<String, dynamic> json) {
    return SubmissionListItem(
      id: json['id'],
      programId: json['program_id'],
      userId: json['user_id'],
      title: json['title'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
      programName: json['program_name'] ?? '',
      studentName: json['student_name'] ?? '',
      studentEmail: json['student_email'] ?? '',
      messageCount: json['message_count'] ?? 0,
      unreadCount: json['unread_count'] ?? 0,
      lastMessageAt: DateTime.parse(json['last_message_at']),
      lastMessageText: json['last_message_text'] ?? '',
      lastMessageFrom: json['last_message_from'] ?? '',
    );
  }
}

class UnreadCounts {
  final int total;
  final Map<String, int> byProgram;
  final Map<String, int> bySubmission;

  UnreadCounts({
    required this.total,
    required this.byProgram,
    required this.bySubmission,
  });

  factory UnreadCounts.fromJson(Map<String, dynamic> json) {
    return UnreadCounts(
      total: json['total'] ?? 0,
      byProgram: Map<String, int>.from(json['by_program'] ?? {}),
      bySubmission: Map<String, int>.from(json['by_submission'] ?? {}),
    );
  }
}
